### DISCUSSION TITLE
Difficulty when porting WPF control to Avalonia

### DISCUSSION THREAD

**Discussion by mykolamor** [AvaloniaUI/Avalonia#5470] [Q&A] [answered]
Hi! I'm currently rewriting [this WPF HexEditor](https://github.com/abbaye/WpfHexEditorControl) to Avalonia and I have some questions about the performance of different Avalonia functions and about my assumptions for replacing WPF functions with Avalonia ones. I am not very familiar with Avalonia so I may have made some mistakes.

Here are the changes I came up with:
- `Loaded` event to `AttachedToVisualTree` because the handler uses `Width` which should be populated after being attached?
- `ActualWidth` to `Width`.
- `SizeChangedEvent` to a `Bounds` observable.
- `MouseWheelEvent` to `PointerWheelChangedEvent`.
- `MouseDownEvent` to `PointerPressedEvent`.
- `DropEvent` to `DragDrop.DropEvent`.

Here's what I don't know:
1. What is better to use to register StyledProperty.Changed handlers:
```
SelectionFirstColorProperty.Changed.AddClassHandler<HexEditor>((x, e) => Control_ColorPropertyChanged(x, e));
SelectionSecondColorProperty.Changed.AddClassHandler<HexEditor>((x, e) => Control_ColorPropertyChanged(x, e));
```
or
```
Observable.Merge(
    this.GetPropertyChangedObservable(SelectionFirstColorProperty),
    this.GetPropertyChangedObservable(SelectionSecondColorProperty)
).Subscribe(x => Control_ColorPropertyChanged(this, x));
```
or
```
this.GetPropertyChangedObservable(SelectionFirstColorProperty).Subscribe(x => Control_ColorPropertyChanged(this, x));
this.GetPropertyChangedObservable(SelectionSecondColorProperty).Subscribe(x => Control_ColorPropertyChanged(this, x));
```

2. What is better to use to register a Click handler:
`UndoCMenu.Click += UndoCMenu_Click;`
or
`UndoCMenu.AddHandler(MenuItem.ClickEvent, UndoCMenu_Click);`

3. Is there a way to use the property changed signature for validation? I have:
```
public static readonly StyledProperty<bool> IsStringDataVisibleProperty =
    AvaloniaProperty.Register<HexEditor, bool>(
        nameof(IsStringDataVisible),
        defaultValue: true,
        coerce: Visibility_CoerceValue,
        validate: IsStringDataVisible_ValidateValue
    );
```
but the validation method in WPF [is defined as](https://github.com/abbaye/WpfHexEditorControl/blob/9e75135197e56afaf457dba8c985c50b7a2428c3/Sources/WPFHexaEditor/HexEditor.xaml.cs#L1890)
```
private static void StringDataVisibility_ValidateValue(DependencyObject d, DependencyPropertyChangedEventArgs e)
```
and it is crucial for the logic inside this method to have both `d` and `e`. But in Avalonia, validation is done through `Func<TValue, bool>` which is not suitable in this scenario. How do I overcome this?

4. Lastly, I came across the `GetCharFromKey` function, which takes a key from the enum and outputs a char from the virtual keyboard. It can be found [here](https://github.com/abbaye/WpfHexEditorControl/blob/731249a43944fa610704e3c542c956dc333253fb/Sources/WPFHexaEditor/Core/KeyValidator.cs#L149). I understand that the code is platform-specific, but is there an existing cross-platform way to implement the method in Avalonia at least on desktop?

I know that I'm asking a lot of questions at once, but getting at least some help would be priceless.

**Comment by jp2masa** [comment, marked as answer]
`ActualWidth` corresponds to `Bounds.Width`.

1. `AddClassHandler` (in the static constructor) is the best way, because it only subscribes once for all control instances.
2. They're actually equivalent:
https://github.com/AvaloniaUI/Avalonia/blob/6332dfca96430e45cf61a82d28d5765358175d1a/src/Avalonia.Controls/MenuItem.cs#L162
3. I don't think that's validate, if I understand correctly the [original](https://github.com/abbaye/WpfHexEditorControl/blob/731249a43944fa610704e3c542c956dc333253fb/Sources/WPFHexaEditor/HexEditor.xaml.cs#L1886-L1888) is calling this constructor: [FrameworkPropertyMetadata](https://docs.microsoft.com/en-us/dotnet/api/system.windows.frameworkpropertymetadata.-ctor?view=net-5.0#System_Windows_FrameworkPropertyMetadata__ctor_System_Object_System_Windows_PropertyChangedCallback_System_Windows_CoerceValueCallback_), so it would be equivalent to a property changed subscription like 1. (`AddClassHandler` or `StringDataVisibilityProperty.Changed.Subscribe`) .
4. That would be something like #4447, if I understand correctly? If so, I think there's currently no way of doing that.

**Reply by mykolamor** [reply]
Thanks for all the help! I implemented all of the suggestions and everything seems OK for now. The #4447 not implemented is a shame though, I think many projects would benefit from that, like game editor UI etc.
